packages/financial-templates-lib/src/price-feed/utils.js

```diff
@@ -168,6 +168,97 @@ exports.PriceHistory = (getPrice, prices = {}) => {
   };
 };
 
+/**
+ * @notice Constructs the BlockFinder, which is a simple class-like object to find blocks by timestamp, but keeps a
+ *      cache to optimize the search.
+ * @param {Function} requestBlock async function, like web3.eth.getBlock that returns a block for a particular block
+ *      number or returns the latest block if no argument is provided. Blocks returned must have a `number` and
+ */
+exports.BlockFinder = (requestBlock, blocks = []) => {
+  // Grabs the most recent block and caches it.
+  async function getLatestBlock() {
+    const block = await requestBlock();
+    const index = lodash.sortedIndexBy(blocks, block, "number");
+    if (blocks[index]?.number !== block.number) blocks.splice(index, 0, block);
+    return blocks[index];
+  }
+
+  // Grabs the block for a particular number and caches it.
+  async function getBlock(number) {
+    const index = lodash.sortedIndexBy(blocks, { number }, "number");
+    if (blocks[index]?.number === number) return blocks[index]; // Return early if block already exists.
+    const block = await requestBlock(number);
+    blocks.splice(index, 0, block); // A simple insert at index.
+    return block;
+  }
+
+  // Effectively, this is an interpolation search algorithm to minimize block requests.
+  async function findBlock(startBlock, endBlock, timestamp) {
```

Comment:
do you want to apply any input sanitization on these variables or types?

Reply:
This is a great question. Only one function is exported, so the others aren't callable from outside. If these methods are only called internally, should we do input sanitization? Or should we only do it for the externally-callable functions?

Reply:
@chrismaree I added some input validation both to the outer function (BlockFinder) and the getBlockForTimestamp.